experiments/process_sandbox/include/process_sandbox/sandbox.h

```diff
@@ -20,95 +21,383 @@
+       */
+      void* base;
+
+      /**
+       * Top of the shared memory region.
+       */
+      void* top;
+
+      /**
+       * Chunks that have been allocated by the shared allocator and not yet
+       * returned to the child.
+       */
+      snmalloc::ChunkAllocatorState chunk_alloc_state;
+
+    public:
+      /**
+       * The address-space manager that manages this region.
+       */
+      snmalloc::AddressSpaceManager<snmalloc::PALNoAlloc<snmalloc::DefaultPal>>
+        shared_asm;
+
+      /**
+       * Constructor.  Takes the memory range allocated for the sandbox heap as
+       * arguments.  This class takes responsibility for allocating memory from
+       * the provided range.  Nothing should access any of the memory in this
+       * range without first calling the `reserve` method on this class to
+       * acquire a chunk.
+       */
+      LocalState(void* start, size_t size);
+
+      /**
+       * Returns the chunk allocator state for this sandbox.  This is called
+       * only by the corresponding method in the `SharedAllocConfig` class,
+       * which is called from snmalloc.
+       */
+      snmalloc::ChunkAllocatorState& get_slab_allocator_state()
+      {
+        return chunk_alloc_state;
+      }
+
+      /**
+       * Predicate to test whether an object of size `sz` starting at `ptr`
+       * is within the region managed by this memory provider.
+       */
+      bool contains(const void* ptr, size_t sz)
+      {
+        // We shouldn't need the const cast here, but pointer_offset doesn't
+        // correctly handle const pointers yet.
+        return (ptr >= base) &&
+          (pointer_offset(const_cast<void*>(ptr), sz) < top);
+      }
+
+      /**
+       * Return the top of the sandbox.
+       */
+      void* top_address()
+      {
+        return top;
+      }
+
+      /**
+       * Return the top of the sandbox.
+       */
+      void* get_base()
+      {
+        return base;
+      }
+    };
 
-  public:
     /**
-     * Constructor.  Takes the memory range allocated for the sandbox heap as
-     * arguments.
+     * The type of a pagemap that spans the entire address space.
+     *
+     * This is used for all compartments.  Currently, this is not the same
+     * pagemap that the parent process uses for any non-shared allocations.
+     * This could be changed in environments (such as Verona) that can
+     * guarantee that in-compartment pointers are never freed with the global
+     * `free` or if we have a default `Alloc` that checks for this case
+     * (perhaps with a flag in the metadata entry indicating whether the slab
+     * is compartment-owned).
+     */
+    class Pagemap
+    {
+      /**
+       * Concrete instance of a pagemap.  This is updated only with
+       * `pagemap_lock` held.
+       */
+      inline static snmalloc::FlatPagemap<
+        snmalloc::MIN_CHUNK_BITS,
+        snmalloc::MetaEntry,
+        NoOpPal,
+        /*fixed range*/ false>
```

Comment:
Maybe a comment here to clarify that this pagemap does have a fixed range defined by the base and top, but uses the non-fixed range implementation.

Reply:
Added, thanks.